genai-perf/genai_perf/metrics/telemetry_statistics.py

```diff
@@ -0,0 +1,112 @@
+# PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE COPYRIGHT OWNER OR
+# CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+# EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+# PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+# PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+# OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+from collections import defaultdict
+from typing import Any, DefaultDict, Dict, List
+
+from genai_perf.metrics.statistics_util import StatisticsUtil
+from genai_perf.metrics.telemetry_metrics import TelemetryMetrics
+
+
+class TelemetryStatistics:
+    """A class that aggregates various statistics from telemetry metrics class."""
+
+    def __init__(self, metrics: TelemetryMetrics):
+        self._metrics = metrics
+        self._stats_dict: DefaultDict[str, Any] = defaultdict(lambda: defaultdict(dict))
+
+        self._add_units()
+        for attr, data in self._metrics.data.items():
+            if self._should_skip(data):
+                continue
+            self._calculate_statistics(data, attr)
+
+    def _should_skip(self, data: Dict[str, List[float]]) -> bool:
+        """Checks if some metrics should be skipped."""
+        if len(data) == 0:
+            return True
+        return False
+
+    def _calculate_statistics(self, data: Dict[str, List[float]], attr: str) -> None:
+        for gpu_index, gpu_data in data.items():
+            self._calculate_mean(gpu_data, attr, gpu_index)
+            if not self._is_constant_metric(attr):
+                self._calculate_percentiles(gpu_data, attr, gpu_index)
+                self._calculate_minmax(gpu_data, attr, gpu_index)
+                self._calculate_std(gpu_data, attr, gpu_index)
+
+    def _calculate_mean(self, data: List[float], attr: str, gpu_index: str) -> None:
+        avg = StatisticsUtil.calculate_mean(data)
+        setattr(self, f"avg_{attr}_{gpu_index}", avg)
+        self._stats_dict[attr][gpu_index]["avg"] = avg
+
+    def _calculate_percentiles(
+        self, data: List[float], attr: str, gpu_index: str
+    ) -> None:
+        percentile_results = StatisticsUtil.calculate_percentiles(data)
+        p25 = percentile_results[0]
+        p50 = percentile_results[1]
+        p75 = percentile_results[2]
+        p90 = percentile_results[3]
+        p95 = percentile_results[4]
+        p99 = percentile_results[5]
+        setattr(self, f"p25_{attr}_{gpu_index}", p25)
+        setattr(self, f"p50_{attr}_{gpu_index}", p50)
+        setattr(self, f"p75_{attr}_{gpu_index}", p75)
+        setattr(self, f"p90_{attr}_{gpu_index}", p90)
+        setattr(self, f"p95_{attr}_{gpu_index}", p95)
+        setattr(self, f"p99_{attr}_{gpu_index}", p99)
+        self._stats_dict[attr][gpu_index]["p99"] = p99
```

Comment:
One quick idea that popped on my head, what do you think about restructuring the `stats_dict` into:
```python
self._stats_dict[gpu_index][attr]["p99"]
```
so that you index for gpu id first and then for attribute? If this restructuring is possible, then I think we can reuse `Statistics` method (although it would require a small refactor, but I think it's possible).

Reply:
Restructuring as self._stats_dict[gpu_index][attr]["p99"] is possible. I see what you mean.
Yeah.. I guess it can work. Thanks for the suggestion @nv-hwoo

Reply:
I modified as per your suggestion and our offline discussion.
I made Statistics and TelemetryStatistics to have composition instead of Inheritance relationship as that made more sense with different metric types. Also, not calling super.__init__() while using inheritance gives a codeql error.
I was able to modify without changing the self._stats_dict structure in TelemetryStatistics.